### PR TITLE
[Security Solution][Detections]Rules _bulk_action edit API

### PR DESCRIPTION
**Pull request by vitaliidm**
Adds edit action to rules _bulk_action API

Adresses: https://github.com/elastic/kibana/pull/120472

Preview [here](https://security-docs_1317.docs-preview.app.elstc.co/guide/en/security/master/bulk-actions-rules-api.html).

### CODE REVIEW

**Comment by banderror** [comment]
Just a note to not forget to document that the order of edits specified in the request body will matter.
https://github.com/elastic/kibana/pull/120472#discussion_r773767229

**Comment by jmikell821** [comment]
Hey @vitaliidm - just checking on the status of this. Does this apply to `8.0+`?

**Comment by vitaliidm** [comment]
Hi, @jmikell821 , sorry for the late response. Yes, we are shipping it in 8.1. 
I have updated this PR with all recent changes to API and it's ready to review.
Let me know please, if I need to do anything else

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 1]
```diff
@@ -255,15 +308,247 @@ POST api/detection_engine/rules/_bulk_action
 [discrete]
 ===== Response payload
 
-For `enable`, `disable`, `delete`, and `duplicate` actions, a JSON object containing the outcome of the call and the number of affected rules:
+For `enable`, `disable`, `delete`, `edit`, and `duplicate` actions, a JSON object containing the outcome of the action: number of failed, succeeded, total rules, as well as full rule's objects(either created, deleted or updated depends on action) in array.
```
Comment (benironside) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:311]:
```suggestion
For `enable`, `disable`, `delete`, `edit`, and `duplicate` actions, a JSON object containing the action's outcome: number of failed and successful outcomes, number of rules, and number of rule objects created, deleted or updated (depending on the action).
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 2]
```diff
@@ -255,15 +308,247 @@ POST api/detection_engine/rules/_bulk_action
...
 }
 --------------------------------------------------
 
 
 For an `export` action, an `.ndjson` file containing exported rules.
+
+
+*Example 2, Partial failure*
+
+The following request will add tags `tag-1`, `tag-2` to rules, which ids are send in payload:
```
Comment (benironside) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:406]:
```suggestion
The following request adds tags `tag-1` and `tag-2` to the rules that have the IDs sent in the payload:
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 3]
```diff
@@ -255,15 +308,247 @@ POST api/detection_engine/rules/_bulk_action
...
+  "action": "edit",
+  "edit": [{ action: "add_tags", value:["tag-1", "tag-2"] }]
+}
+--------------------------------------------------
+
+[discrete]
+===== Response code
+
+`500`::
+    Partial failure of bulk action
```
Comment (benironside) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:424]:
```suggestion
    Indicates partial bulk action failure.
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 4]
```diff
@@ -255,15 +308,247 @@ POST api/detection_engine/rules/_bulk_action
...
+
+[discrete]
+===== Response payload
+
+If processing of any rule fails, partial error will be return with id/name of affected rule and corresponding error together with successfully processed rules(same format as for successful 200 request)
```
Comment (benironside) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:429]:
```suggestion
If processing of any rule fails, a partial error outputs the ID and/or name of the affected rule and the corresponding error, as well as successfully processed rules (in the same format as a successful 200 request).
```

**Review by benironside — COMMENTED** [review]
Left some suggestions, I'll take another look once everyone's had a chance to comment.

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 5]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+| `set_tags` | String[] |  Set value as tags in rules
+| `add_index` | String[] |  Adds index patterns to rules
+| `delete_index` | String[] |  Delete index patterns from rules
+| `set_index` | String[] |  Set index patterns in rules
+| `set_timeline` a| object with 2 properties:
+
+* `timeline_id`: String
+* `timeline_title`: String
+
+ |  Adds timeline to rule
```
Comment (nastasha-solomon) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:281]:
```suggestion
 |  Adds Timeline to rule
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 6]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+|==============================================
+
+Actions(<<bulk-edit-object-schema, BulkEditAction[]>> ) in `edit` array payload's property are applying in a historical order(from first to last).
```
Comment (nastasha-solomon) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:284]:
```suggestion
<<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the`edit` array payload's property.
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 7]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+|`type` field |`value` field |Description
+| `add_tags` | String[] |  Adds tags to rules
+| `delete_tags` | String[] |  Delete tags from rules
+| `set_tags` | String[] |  Set value as tags in rules
+| `add_index` | String[] |  Adds index patterns to rules
+| `delete_index` | String[] |  Delete index patterns from rules
+| `set_index` | String[] |  Set index patterns in rules
+| `set_timeline` a| object with 2 properties:
+
+* `timeline_id`: String
```
[the remark this answers is not in the document]
Comment (jmikell821) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:278]: @vitaliidm or @banderror can you answer this? ^

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 8]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+
+Actions(<<bulk-edit-object-schema, BulkEditAction[]>> ) in `edit` array payload's property are applying in a historical order(from first to last).
 
 [discrete]
-====== Example request
+===== Example requests
+
+*Example 1*
 
 The following request will activate all rules that have the `test` tag:
```
Comment (nastasha-solomon) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:291]:
```suggestion
The following request activates all rules with the `test` tag:
```

**Review by nastasha-solomon — COMMENTED** [review]
Thanks for writing this up @vitaliidm! Just left a handful of edits.

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 9]
```diff
@@ -226,15 +226,69 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+| `set_index` | String[] |  Set index patterns in rules
+| `set_timeline` a| object with 2 properties:
+
+* `timeline_id`: String
+* `timeline_title`: String
+
+ |  Adds Timeline to rule
+|==============================================
+
+<<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the`edit` array payload's property.
```
Comment (joepeeples) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:284]: Missing space
```suggestion
<<bulk-edit-object-schema, Actions>> are shown in order of oldest to newest in the `edit` array payload's property.
```

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 10]
```diff
@@ -216,7 +216,7 @@ A JSON array containing the updated rules.
 [[bulk-actions-rules-api-action]]
 ==== Bulk action
 
-Applies a bulk action to multiple rules. The bulk action is applied to all rules that match the filter.
+Applies a bulk action to multiple rules. The bulk action is applied to all rules that match the filter or to the list of rules by their ids.
```
Comment (joepeeples) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:219]:
```suggestion
Applies a bulk action to multiple rules. The bulk action is applied to all rules that match the filter or to the list of rules by their IDs.
```

**Review by joepeeples — COMMENTED** [review]
The other reviewers already caught most things, but I added a few more edits.

Also more generally, I'm wondering if we could help distinguish the overall structure of the page, with so many section and subsection header levels. Maybe a simple horizontal line  between major sections? (Syntax is `'''` per [Asciidoc spec](https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#breaks).)

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 11]
```diff
@@ -216,7 +216,7 @@ A JSON array containing the updated rules.
 [[bulk-actions-rules-api-action]]
 ==== Bulk action
```
Comment (spong) [inline docs/detections/api/rules/rules-api-bulk-actions.asciidoc:217]: This is an existing page question, but any reason `Create`/`Update` have the `this API supports Token-based authentication` warning 
<p align="center">
   
</p>



but `Delete`/`Action` do not?
<p align="center">
   
</p>

It's the same for the whole API, so can this just be a page-level warning?

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 12]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+
+[discrete]
+===== Possible `BulkEditAction` object values
+
+[width="100%",options="header"]
+|==============================================
+|`type` field |`value` field |Description
+| `add_tags` | String[] |  Adds tags to rules
+| `delete_tags` | String[] |  Delete tags from rules
+| `set_tags` | String[] |  Set value as tags in rules
```
[the remark this answers is not in the document]
Comment (spong) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:272]: May want additional clarification here as the `set_` action's will overwrite the field with whatever `value` is provided. This aligns with `Set value as tags in rules` but not `Set index patterns in rules` -- may want to align all three?

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (spong) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:278]: So these two fields are actually part of the existing Rule CRUD API, just currently undocumented (they could be listed in the docs for [optional fields for all rule types](https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all), but we should verify as to why they weren't included initially). For the existing CRUD API these fields are optional ([source](https://github.com/elastic/kibana/blob/ba0833f42b89a0b62bccd1d8c04bf9c799f32258/x-pack/plugins/security_solution/common/detection_engine/schemas/request/rule_schemas.ts#L153-L154)), and so providing _either_ a `timeline_id` or `timeline_title` should validate.

As for the bulk edit API above though, it appears as if both are required ([source](https://github.com/elastic/kibana/blob/1f3f1c6fc6f8b4b8036c92161854b10814758708/x-pack/plugins/security_solution/common/detection_engine/schemas/common/schemas.ts#L462-L465)). 

For context, these are the same `id`/`name` fields used within the [Timeline API](https://www.elastic.co/guide/en/security/current/_get_timeline_template_by_templatetimelineid.html). As far as I can remember these fields are only being used for the `Investigate in Timeline` action on Alerts, and results in that specific alert data being opened in the specified timeline template (by means of [timelineId](https://github.com/elastic/kibana/blob/e312c36e4cf67dad48a58fdcd947eab022934315/x-pack/plugins/security_solution/public/detections/components/alerts_table/actions.tsx#L520)).

Hopefully that provides some context here, but let me know if you have anymore questions! We should discuss as a team if these fields can be documented/supported for the Rules API as a whole @banderror.

**Review by spong — APPROVED** [review]
Reviewed taking into account the existing suggestions and overall LGTM! Couple nits, and we need to determine if we're going to expose the timeline id/name fields throughout the rest of the Rule API's, but further team discussion may be required on that one.

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc (hunk 7 above)
[the remark this answers is not in the document]
Comment (spong) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:278]: I forgot we don't expose a `Bulk Update Timeline` UI, so with these fields not being documented in the other Rule API's perhaps this doesn't need to be documented here just yet either?
<p align="center">
  
</p>

[the remark this answers is not in the document]
Comment (vitaliidm) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:278]: Sounds reasonable to me.
I'm removing timeline parameter for now, until we add UI functionality

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc [hunk 13]
```diff
@@ -226,13 +226,67 @@ Applies a bulk action to multiple rules. The bulk action is applied to all rules
...
+
+Yes, if action is `edit`
+|==============================================
+
+
+[[bulk-edit-object-schema]]
+[discrete]
+==== BulkEditAction object
+
+* `type` - enum - that defines what will be updated in rules
```
[the remark this answers is not in the document]
Comment (vitaliidm) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:261]: Screenshot 2022-03-03 at 12 38 40

this enum is already incorporated in table

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc (hunk 12 above)
[the remark this answers is not in the document]
Comment (vitaliidm) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:272]: Agree with @spong . I rewritten it as

- Overwrite tags with value in rules
- Overwrite index patterns with value in rules

**Comment by vitaliidm** [comment]
> Also more generally, I'm wondering if we could help distinguish the overall structure of the page, with so many section and subsection header levels. Maybe a simple horizontal line between major sections? (Syntax is ''' per [Asciidoc spec](https://docs.asciidoctor.org/asciidoc/latest/syntax-quick-reference/#breaks).)

Thanks for the suggestion @joepeeples , I've split APIs section as you advised 👍 

@joepeeples, @benironside, @nastasha-solomon I've applied suggested changes. Please have a look and let me know if you happy with merging this PR. Thank you

File: docs/detections/api/rules/rules-api-bulk-actions.asciidoc (hunk 11 above)
Comment (vitaliidm) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:225]: @spong, not sure about this. Is there a way to determine which type of authentication is supported by particular API(maybe looking into route configuration)? 

Take for example this one, read rule https://www.elastic.co/guide/en/security/current/rules-api-get.html
and create rule https://www.elastic.co/guide/en/security/current/rules-api-create.html#opt-fields-all

Create API has warning, but by looking into route configuration I don't see the difference. We might want to align these warnings to be consistent across different APIs

Comment (spong) [reply docs/detections/api/rules/rules-api-bulk-actions.asciidoc:225]: > @spong, not sure about this. Is there a way to determine which type of authentication is supported by particular API(maybe looking into route configuration)?

I believe so long as security is enabled (which it is by default now) all routes will go through this auth ([source](https://github.com/elastic/kibana/blob/6627bd8b3ab49415e397495d8e076f9861fcc894/x-pack/plugins/security/server/authentication/providers/token.ts#L101-L106)).

So yeah, I think it's just a consistency thing -- generally, auth shouldn't vary between these Kibana CRUD routes, so I think we just need to align on these warnings like you said.

**Review by nastasha-solomon — APPROVED** [review]

**Pull request merged**